LIPS/lip-25.md

```diff
@@ -0,0 +1,1983 @@
+
+struct StakingModule {
+    /// @notice unique id of the staking module
+    uint24 id;
+    /// @notice address of staking module
+    address stakingModuleAddress;
+    /// @notice part of the fee taken from staking rewards that goes to the staking module
+    uint16 stakingModuleFee;
+    /// @notice part of the fee taken from staking rewards that goes to the treasury
+    uint16 treasuryFee;
+    /// @notice maximum stake share that can be allocated to a module, in BP
+    uint16 stakeShareLimit; // formerly known as `targetShare`
+    /// @notice staking module status if staking module can not accept the deposits or can participate in further reward distribution
+    uint8 status;
+    /// @notice name of staking module
+    string name;
+    /// @notice block.timestamp of the last deposit of the staking module
+    /// @dev NB: lastDepositAt gets updated even if the deposit value was 0 and no actual deposit happened
+    uint64 lastDepositAt;
+    /// @notice block.number of the last deposit of the staking module
+    /// @dev NB: lastDepositBlock gets updated even if the deposit value was 0 and no actual deposit happened
+    uint256 lastDepositBlock;
+    /// @notice number of exited validators
+    uint256 exitedValidatorsCount;
+    /// @notice module's share threshold, upon crossing which, exits of validators from the module will be prioritized, in BP
+    uint16 priorityExitShareThreshold;
+    /// @notice the maximum number of validators that can be deposited in a single block
+    /// @dev must be harmonized with `OracleReportSanityChecker.churnValidatorsPerDayLimit`
+    /// (see docs for the `OracleReportSanityChecker.setChurnValidatorsPerDayLimit` function)
+    uint64 maxDepositsPerBlock;
+    /// @notice the minimum distance between deposits in blocks
+    /// @dev must be harmonized with `OracleReportSanityChecker.churnValidatorsPerDayLimit`
+    /// (see docs for the `OracleReportSanityChecker.setChurnValidatorsPerDayLimit` function)
+    uint64 minDepositBlockDistance;
+}
+
+struct StakingModuleCache {
+    address stakingModuleAddress;
+    uint24 stakingModuleId;
+    uint16 stakingModuleFee;
+    uint16 treasuryFee;
+    uint16 stakeShareLimit;
+    StakingModuleStatus status;
+    uint256 activeValidatorsCount;
+    uint256 availableValidatorsCount;
+}
+
+/**
+* @notice register a new staking module
+* @param _name name of staking module
+* @param _stakingModuleAddress address of staking module
+* @param _stakeShareLimit maximum share that can be allocated to a module
+* @param _priorityExitShareThreshold module's proirity exit share threshold
+* @param _stakingModuleFee fee of the staking module taken from the consensus layer rewards
+* @param _treasuryFee treasury fee
+* @param _maxDepositsPerBlock the maximum number of validators that can be deposited in a single block
+* @param _minDepositBlockDistance the minimum distance between deposits in blocks
+*/
+function addStakingModule(
```

Comment:
Consider leaving only the names (or interfaces) of the external (!) methods that are changed and more explanation in words how it should work. Now quite a lot of code is in LIP and only a part of it changes. Code inserts could probably be useful somewhere to expose individual thoughts, but bringing absolutely every change to the LIP seems unnecessary

Reply:
Yes, in general, I agree with this suggestion. The text of the LIP has been substantially rewritten. Many text explanations that cover the logic of the implemented code have been added. Many internal methods and structures with not very important updates have been removed from code examples.